File: 661_image_smoother/image_smoother.cpp
/*
 * @Date: 2022-03-23 23:29:21
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-03-23 23:34:46
 * @FilePath: /algorithm/661_image_smoother/image_smoother.cpp
 */

#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<vector<int>> imageSmoother(vector<vector<int>> img) {
    int m = img.size(), n = img[0].size();
    vector<vector<int>> ret(m, vector<int>(n));
    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        int num = 0, sum = 0;
        for (int x = i - 1; x <= i + 1; x++) {
          for (int y = j - 1; y <= j + 1; y++) {
            if (x >= 0 && x < m && y >= 0 && y < n) {
              num++;
              sum += img[x][y];
            }
          }
        }
        ret[i][j] = sum / num;
      }
    }
    return ret;
  }
};

int main() {
  assert((Solution().imageSmoother(
              vector<vector<int>>{{1, 1, 1}, {1, 0, 1}, {1, 1, 1}}) ==
          vector<vector<int>>{{0, 0, 0}, {0, 0, 0}, {0, 0, 0}}));
  assert(
      (Solution().imageSmoother(vector<vector<int>>{
           {100, 200, 100}, {200, 50, 200}, {100, 200, 100}}) ==
       vector<vector<int>>{{137, 141, 137}, {141, 138, 141}, {137, 141, 137}}));
}

File: golang/661_image_smoother/image_smoother.go
/*
 * @Date: 2022-03-23 23:29:28
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-03-23 23:41:28
 * @FilePath: /algorithm/661_image_smoother/image_smoother.go
 */

package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func imageSmoother(img [][]int) [][]int {
	min := func(a, b int) int {
		if a > b {
			return b
		}
		return a
	}
	max := func(a, b int) int {
		if b > a {
			return b
		}
		return a
	}
	m, n := len(img), len(img[0])
	ans := make([][]int, m)
	for i := range ans {
		ans[i] = make([]int, n)
		for j := range ans[i] {
			sum, num := 0, 0
			for _, row := range img[max(i-1, 0):min(i+2, m)] {
				for _, v := range row[max(j-1, 0):min(j+2, n)] {
					sum += v
					num++
				}
			}
			ans[i][j] = sum / num
		}
	}
	return ans
}

func main() {
	tests := []struct {
		img [][]int
		ans [][]int
	}{
		{[][]int{{1, 1, 1}, {1, 0, 1}, {1, 1, 1}}, [][]int{{0, 0, 0}, {0, 0, 0}, {0, 0, 0}}},
		{[][]int{{100, 200, 100}, {200, 50, 200}, {100, 200, 100}}, [][]int{{137, 141, 137}, {141, 138, 141}, {137, 141, 137}}},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, imageSmoother(test.img), index)
	}
}
